design/system_tests.md

```diff
@@ -0,0 +1,42 @@
+# System Tests
+
+## Requirements
+1. Test the system in the same manner users will be using it. 
+   No ad-hoc initializations or mocking of any kind. 
+2. Cover the most common use-cases.
+3. Portability: The tests don't assume anything on the environment where the lakeFS & DB are running.
+   lakeFS address is the only **required** argument for running the tests.  
+
+## Non-Requirements
+1. Test every use-case of the system, including:
+    a. Every API/gateway endpoint.
+    b. Every application (e.g. Spark) usage pattern.
+2. Avoid repeating unit/integration tests scenarios.
+3. Stress testing / benchmarks. 
+
+## Solution
+### Test
+A test consists of one or more calls to the API/gateway and assertions on the actual output.
+Test is a standalone unit. It shouldn't rely on any other test to pass to start execution. 
+
+Later on we might consider adding checks on the DB or data lake layer(S3),
```

Comment:
There is some feature that probably will require some sort of looking under the hood - retention, and duplication for example.
Currently, we can check duplication by checking our catalog physical address - but the real test is also to check that the original file we uploaded is gone.

Reply:
"No...", but really "yes!!!".  Retention doesn't actually delete anything, it only tags the files for deletion.  But we _can_ test for those tags on all relevant objects, so yes.  A tougher issue is how to to end up with files to expire in the first place: we don't allow expiry periods shorter than 2 hours, and even that is not really configurable.  I am hesitant to add a flag to remove this safety measure from lakeFS.
None of this should block proceeding with other system tests.  Also I think _how_ to test retention is out of scope for a design of the _general_ way to test things: these issues are somewhat specific.

Reply:
I don't consider retention and deduplication as P0 for these tests, first tests will test just the core functionality.